### DISCUSSION TITLE
Byte Battle Contract

### DISCUSSION THREAD

**Discussion by Ryang-21** [stellar/sorobanathon#41] [General]
# Byte Battle
This contract is a simple game that allows two users to place a wager and put their public keys to battle. The contract determines which u8 in the public keys that will be compared by modding the timestamp by 32. The winner of the battle then receives the other's wager. 
The repo can be found here https://github.com/Ryang-21/Soroban-Byte-Battle

## Agreed Upon Matchmaking
Soroban's configurable authentication system allows for contracts to determine how they would like to authenticate calls to the contract. This allows for interesting authentication schemes such as agreed upon matchmaking. The battle function requires signatures from both players which contain their own identifier, the opponents identifier, and the wager amount. Requiring both players identifiers and the wager amount in each signature ensures that each player has agreed to play the other for the specified wager amount. Replay prevention is provided through the nonce and is incremented for each player when a successful battle occurs. 

## Implementation
The authentication of the signatures is performed by calling verify on each signature and providing the following arguments
1. Environment 
2. The provided player signature
3. The name of the function as a symbol
4. A tuple of arguments consisting of the identifier of the signer, signers nonce, the opponents identifier, and the wager amount

```rust
fn battle(
        e: Env,
        player1_sig: Signature,
        player2_sig: Signature,
        amount: BigInt,
    ) -> (Bytes, Bytes, u32) {
        let signer1_id = player1_sig.identifier(&e);
        let signer2_id = player2_sig.identifier(&e);
        let p1_nonce = get_nonce(&e, &signer1_id);
        let p2_nonce = get_nonce(&e, &signer2_id);

        verify(
            &e,
            &player1_sig,
            symbol!("battle"),
            (&signer1_id, &p1_nonce, &signer2_id, &amount),
        );
        verify_and_consume_nonce(&e, &player1_sig, &p1_nonce);
        verify(
            &e,
            &player2_sig,
            symbol!("battle"),
            (&signer2_id, &p2_nonce, &signer1_id, &amount),
        );
        verify_and_consume_nonce(&e, &player2_sig, &p2_nonce);
```
<br>
The players public keys were obtained by using rust's match function to match the players’ identifer against the three types of identifers. This allows the public key to be extracted from the identifer. The account identifer was a special case as it does not contain a BytesN<32> and instead has an Accountid. To convert the account id into a bytes object the sdk provides a serialize function that returns the account id as bytes. A unique result from serializing the account id was that the byte object contained 48 u8 numbers with the first 16 u8 numbers containing mostly 0. This required each player to have their own byte_to_compare number that could then be incrementented by 16 if an account identifer was being used. 

```rust
        let mut player1_byte_to_compare = (e.ledger().timestamp() % 32) as u32;
        let mut player2_byte_to_compare = (e.ledger().timestamp() % 32) as u32;
        let player1_key: Bytes = match signer1_id.clone() {
            Identifier::Account(a) => {
                player1_byte_to_compare += 16;
                a.serialize(&e)
            }
            Identifier::Contract(k) => Bytes::from_array(&e, &k.to_array()),
            Identifier::Ed25519(k) => Bytes::from_array(&e, &k.to_array()),
        };
        let player2_key: Bytes = match signer1_id.clone() {
            Identifier::Account(a) => {
                player2_byte_to_compare += 16;
                a.serialize(&e)
            }
            Identifier::Contract(k) => Bytes::from_array(&e, &k.to_array()),
            Identifier::Ed25519(k) => Bytes::from_array(&e, &k.to_array()),
        };
```

**Comment by achallagundla** [comment]
WOHOO we got a simple bet-waging contract!  Appreciate your submission, looking forward to reviewing it!

**Comment by achallagundla** [comment]
Heya @Ryang-21  ,
Congratulations! Your contribution to Sorobanathon: First Light has been accepted! 🎉
Your submission may qualify for a reward of 1000 XLM! One last step: to receive that reward, please fill out this [form](https://stellarform.typeform.com/Sorobanathon) completely and accurately.

Please let us know if you have any questions. (You can also message me on discord @ anuxhya)

All the best,
Anuhya

PS: No need to submit the form if you already have, just update me on discord if your wallet address has changed.
